### PR TITLE
List buckets creation

### PR DESCRIPTION
**Pull request by mounika-nalla**
### Description

These changes adds the ability to pass a list of buckets to be created in a cos instance and an existing instance.
 https://github.com/terraform-ibm-modules/terraform-ibm-cos/issues/239
https://github.ibm.com/GoldenEye/issues/issues/4026

#### No release required

- [x] Examples or tests (addition or updates of examples or tests)
- [ ] Documentation update
- [ ] CI-related update (pipeline, etc.)
- [ ] Other changes that don't affect Terraform code

#### Release required

- [ ] Bug fix (patch release (`x.x.X`): Change that fixes an issue and is compatible with earlier versions)
- [x] New feature (minor release (`x.X.x`): Change that adds functionality and is compatible with earlier versions)
- [ ] Breaking change (major release (`X.x.x`): Change that is likely incompatible with previous versions)

##### Release notes content

Introduces a new feature, implemented as a submodule. Features of the root module are unchanged. The buckets submodule takes a map of bucket definitions and calls the root module to create each one. It returns a list of buckets created, including names, crns and endpoints.

---

### Checklist for reviewers

- [x] If relevant, a test for the change is included or updated with this PR.
- [x] If relevant, documentation for the change is included or updated with this PR.

### Merge actions for mergers

- Merge by using "Squash and merge".
- Use a relevant [conventional commit](https://www.conventionalcommits.org/) message that is based on the PR contents and any release notes provided by the PR author.

    The commit message determines whether a new version of the module is needed, and if so, which semver increment to use (major, minor, or patch).

### CODE REVIEW

File: examples/replication/outputs.tf
Comment (shemau) [inline examples/replication/outputs.tf:3]: To get from bucket_names to a bucket_name, perhaps it should be bucket_names[0]. This would return a string rather than an array containing a string.

File: examples/replication/outputs.tf
Comment (shemau) [inline examples/replication/outputs.tf:8]: To get from bucket_names to a bucket_name, perhaps it should be bucket_names[0]. This would return a string rather than an array containing a string.

File: main.tf
Comment (shemau) [reply main.tf:224]: I think we usually require three properties for a bucket, the instance id (or guid, not sure which), the name and either the public or private endpoint.
So I think we should look at where buckets are consumed, maybe the replication example; and add either id or guid to the map so that consumers of the module get back complete bucket references that can be passed around modules and consumed easily.

**Review by shemau — CHANGES_REQUESTED** [review]
The last resource in main.tf should be actioned since this is a breaking change.

The broader change is that this module is effectively a container for bucket. By this I mean that module had a set of properties that identified a single bucket all in one place and by accessing the module output you have all the properties of a bucket.
Whilst implementing this solution, we should maintain that ease of access and the concept of a bucket object. As such returning a set of lists is not very consumable. It would be better to return a map of buckets, where each entry has all the properties for a single bucket. I think this should include COS instance information (currently part of the module). This would enable consumers to extract and pass individual buckets around as required.

**Review by shemau — COMMENTED** [review]
The majority of the feedback is resolved.

Since this is a breaking change, could the following changes be included as well please:
- main.tf line 301 (resource "null_resource" "deprecation_notice") can be removed along with the deprecate variable service_endpoints.
- also remove variable bucket_endpoint (per #271). by doing so, there will no need to merge that PR or wait until the next breaking change.

**Review by shemau — DISMISSED** [review]
Looks good to me.

**Comment by ocofaigh** [comment]
what is the impact on consumers who are currently consuming this module to create buckets? Have we tested to ensure that the upgrade will not destroy a bucket? I understand its a breaking change because module inputs and outputs are changing, but I think we still need to ensure someone’s COS bucket isnt going to get destroyed on upgrade

**Comment by mounika-nalla** [comment]
@ocofaigh When determining whether an upgrade would destroy any buckets, this is the output returned. The code will generate test-cos-bucket-1 and test-cos-bucket-2 if no buckets are specified, just as it does in the original code from the master branch (complete example). So, only the CBR rules will be deleted and recreated when we do the upgrade, as shown below.

`Terraform used the selected providers to generate the following execution
plan. Resource actions are indicated with the following symbols:
  + create
  - destroy

Terraform will perform the following actions:

  # module.cos_bucket1.module.bucket_cbr_rule[0].ibm_cbr_rule.cbr_rule will be destroyed
  # (because module.cos_bucket1.module.bucket_cbr_rule[0] is not in configuration)
  - resource "ibm_cbr_rule" "cbr_rule" {
      - created_at          = "2023-03-15T21:44:18.000Z" -> null
      - created_by_id       = "IBMid-550007TT57" -> null
      - crn                 = "crn:v1:bluemix:public:context-based-restrictions:global:a/ACCOUNT_ID::rule:e0c38fce921c33684de5e626be9a42af" -> null
      ....
    }

  # module.cos_bucket1.module.bucket_cbr_rule["test-cos-bucket-1"].ibm_cbr_rule.cbr_rule will be created
  + resource "ibm_cbr_rule" "cbr_rule" {
      + created_at          = (known after apply)
      + created_by_id       = (known after apply)
      + crn                 = (known after apply)
      + description         = "sample rule for bucket 1"
      + enforcement_mode    = "report"
      ....
    }

  # module.cos_bucket2.module.bucket_cbr_rule[0].ibm_cbr_rule.cbr_rule will be destroyed
  # (because module.cos_bucket2.module.bucket_cbr_rule[0] is not in configuration)
  - resource "ibm_cbr_rule" "cbr_rule" {
      - created_at          = "2023-03-15T21:44:18.000Z" -> null
      - created_by_id       = "IBMid-550007TT57" -> null
      - crn                 = "crn:v1:bluemix:public:context-based-restrictions:global:a/ACCOUNT_ID::rule:e0c38fce921c33684de5e626be9a3ddc" -> null
      ....
    }

  # module.cos_bucket2.module.bucket_cbr_rule["test-cos-bucket-2"].ibm_cbr_rule.cbr_rule will be created
  + resource "ibm_cbr_rule" "cbr_rule" {
      + created_at          = (known after apply)
      + created_by_id       = (known after apply)
      + crn                 = (known after apply)
      + description         = "sample rule for bucket 2"
      + enforcement_mode    = "report"
      ....
    }

Plan: 2 to add, 0 to change, 2 to destroy.

Changes to Outputs:
  - bucket_name1 = [
      - "test-cos-bucket-1",
    ] -> null
  - bucket_name2 = [
      - "test-cos-bucket-2",
    ] -> null
  + buckets      = {
      + test-cos-bucket-1 = {
          + bucket_crn           = "crn:v1:bluemix:public:cloud-object-storage:global:a/ACCOUNT_ID:9309ec9b-e103-46f5-84c5-344a682eafaa:bucket:test-cos-bucket-1"
          + bucket_id            = "crn:v1:bluemix:public:cloud-object-storage:global:a/ACCOUNT_ID:9309ec9b-e103-46f5-84c5-344a682eafaa:bucket:test-cos-bucket-1:meta:rl:us-south:public"
          + bucket_name          = "test-cos-bucket-1"
          + bucket_storage_class = "standard"
          + cos_instance_guid    = "9309ec9b-e103-46f5-84c5-344a682eafaa"
          + cos_instance_id      = "crn:v1:bluemix:public:cloud-object-storage:global:a/ACCOUNT_ID:9309ec9b-e103-46f5-84c5-344a682eafaa::"
          + key_protect_key_crn  = "crn:v1:bluemix:public:kms:us-south:a/ACCOUNT_ID:acd6591b-6693-4724-b3dc-8c2ce90003d7:key:a4d412af-2ab5-4ddc-ab48-df02c6fa028d"
          + resource_group_id    = "65460e85d7e3445f8abcb7ed2e23bb39"
          + s3_endpoint_private  = "s3.private.us-south.cloud-object-storage.appdomain.cloud"
          + s3_endpoint_public   = "s3.us-south.cloud-object-storage.appdomain.cloud"
        }
      + test-cos-bucket-2 = {
          + bucket_crn           = "crn:v1:bluemix:public:cloud-object-storage:global:a/ACCOUNT_ID:9309ec9b-e103-46f5-84c5-344a682eafaa:bucket:test-cos-bucket-2"
          + bucket_id            = "crn:v1:bluemix:public:cloud-object-storage:global:a/ACCOUNT_ID:9309ec9b-e103-46f5-84c5-344a682eafaa:bucket:test-cos-bucket-2:meta:crl:us:public"
          + bucket_name          = "test-cos-bucket-2"
          + bucket_storage_class = "standard"
          + cos_instance_guid    = "9309ec9b-e103-46f5-84c5-344a682eafaa"
          + cos_instance_id      = "crn:v1:bluemix:public:cloud-object-storage:global:a/ACCOUNT_ID:9309ec9b-e103-46f5-84c5-344a682eafaa::"
          + key_protect_key_crn  = "crn:v1:bluemix:public:kms:us-south:a/ACCOUNT_ID:acd6591b-6693-4724-b3dc-8c2ce90003d7:key:a4d412af-2ab5-4ddc-ab48-df02c6fa028d"
          + resource_group_id    = "65460e85d7e3445f8abcb7ed2e23bb39"
          + s3_endpoint_private  = "s3.private.us.cloud-object-storage.appdomain.cloud"
          + s3_endpoint_public   = "s3.us.cloud-object-storage.appdomain.cloud"
        }
    }
`

**Comment by ocofaigh** [comment]
@daniel-butler-irl Are you happy enough with this before merge?

**Comment by daniel-butler-irl** [comment]
@ocofaigh It might be better to pass a list of buckets ie full bucket configurations, rather than just names. If you have multiple bucket configurations (like I have for FSCloud) then this will not be of any use and we still have to use the old method of provisioning. Not sure what way we want to go here...

**Comment by ocofaigh** [comment]
I think I agree - It's very likely someone will want to create multiple buckets with different configurations. So I think we should take same approach as we do for `worker_pools` in [terraform-ibm-base-ocp-vpc](https://github.com/terraform-ibm-modules/terraform-ibm-base-ocp-vpc/blob/main/variables.tf)
@mounika-nalla @shemau thoughts?

**Comment by daniel-butler-irl** [comment]
Also since this is a breaking change maybe we can rename `existing_key_protect_instance_guid` and `key_protect_key_crn` to something more generic maybe  `existing_kms_instance_guid` and `kms_key_crn`? I will be using these variables to pass in hyper protect details which works its just the names are a bit misleading

**Comment by vburckhardt** [comment]
Could we keep 'bucket_name' in addition to adding 'bucket_names' in the interface, so that we do not end up with a breaking change. Either bucket_name or bucket_names should be set.

File: modules/buckets/README.md
Comment (daniel-butler-irl) [inline modules/buckets/README.md:4]: We don't need the stats section on a submodule

File: README.md
Comment (daniel-butler-irl) [inline README.md:42]: I think this is incorrect, should be using the submodule? And add a little more details about the submodule in the main readme.

**Comment by daniel-butler-irl** [comment]
Are we leaving the existing bucket creation parts in? We should probably do a breaking change and replace it with the new submodule. @ocofaigh @vburckhardt

File: examples/existing-resources/README.md
Comment (SirSpidey) [inline examples/existing-resources/README.md:8]: replace `COS Bucket` with Cloud Object Storage bucket (lowercase b and full name).

Can you combine this with the list item that identifies that you're creating a COS instance? Is there a reason to separate them? 

For example, 
Suggestion
```
- A Cloud Object Storage instance in the resource group and region with buckets that are not encrypted.
```

File: modules/buckets/README.md
Comment (SirSpidey) [inline modules/buckets/README.md:7]: Use unordered list items (`-`) for things that are not sequential.

File: modules/buckets/README.md
Comment (SirSpidey) [inline modules/buckets/README.md:12]: `root module's ability...`

but I prefer to rewrite without anthropomorphizing the module (it really has no abilities) and possession, as in

> The submodule extends the root module by including support for multiple buckets....

File: modules/buckets/README.md
Comment (SirSpidey) [inline modules/buckets/README.md:48]: Remove. If there are known issues they should be in the main readme file.

File: modules/buckets/README.md
Comment (SirSpidey) [inline modules/buckets/README.md:64]: These examples aren't in the submodule, so don't belong here, right?

File: profiles/fscloud/README.md
Comment (SirSpidey) [inline profiles/fscloud/README.md:84]: Not a big deal, but bucket doesn't need to be capitalized here, just above, or throughout the readme files because it's not a proper noun or part of a offering name.

File: README.md
Comment (SirSpidey) [inline README.md:2]: don't capitalize module

**Review by SirSpidey — CHANGES_REQUESTED** [review]
some requested changes and some suggestions.

**Comment by daniel-butler-irl** [comment]
@shemau This repo had been migrated to the new pipeline, to trigger post a comment `/run pipeline` when you think it's ready for test

File: modules/buckets/README.md
Comment (ocofaigh) [reply modules/buckets/README.md:64]: I would remove the examples section (and the hooks metadata so they don't get re-added) from here

Comment (SirSpidey) [reply modules/buckets/README.md:62]: Looks like the examples other docs hooks are still in the readme.

**Review by SirSpidey — COMMENTED** [review]
Apart from the pre-commit hooks in the example readme, ~~I think the content is ready to go in this PR~~.

Just looked closer. I think there are some outstanding issues left. I've clicked resolve on the ones that are closed, I think.

**Review by SirSpidey — COMMENTED** [review]
Apart from the pre-commit hooks in the example readme, I think the content is ready to go in this PR.

File: examples/existing-resources/README.md
Comment (SirSpidey) [reply examples/existing-resources/README.md]: Looks like this comment is still open

File: examples/existing-resources/README.md
Comment (SirSpidey) [inline examples/existing-resources/README.md:10]:
```suggestion
- An IAM access policy to allow Cloud Object Storage to access Key Protect for the encryption key
- A Cloud Object Storage instance with no buckets in a resource group and region.
    This example uses the code in the main `terraform-ibm-cos` module.
- Cloud Object Storage buckets with encryption, versioning and archiving respectively.
    This example uses the code in the `buckets` module.
```

putting the no bucket/bucket items next to each other and putting the explanation in a 2nd paragraph.

File: examples/existing-resources/README.md
Comment (SirSpidey) [inline examples/existing-resources/README.md:9]: Suggestion to follow IBM style

```md
An end-to-end example that creates the following infrastructure:
- A resource group, if one is not passed in.
- A Key Protect instance, key ring, and key in a resource group and region.
- An IAM access policy to allow Cloud Object Storage to access Key Protect for the encryption key.
- A Cloud Object Storage instance with no buckets in a resource group and region.
    This example uses the code in the main `terraform-ibm-cos` module.
- Cloud Object Storage buckets with encryption, versioning and archiving respectively.
```

**Review by SirSpidey — COMMENTED** [review]
one suggestion

**Comment by ocofaigh** [comment]
@shemau @daniel-butler-irl @mounika-nalla This PR is passing the pipeline. Can we get a final review done on it so it can be approved and merged?

File: outputs.tf
Comment (daniel-butler-irl) [inline outputs.tf:34]: Should this be outputting a list of buckets now?

**Review by daniel-butler-irl — CHANGES_REQUESTED** [review]

**Review by daniel-butler-irl — APPROVED** [review]

File: README.md
Comment (SirSpidey) [reply README.md:42]: If you update this, replace `above` with a "non-spatial" work or phrase. We avoid those words to help users who rely on screen readers.

For example, 
> Creates more buckets in the instance created in the first (or earlier) example.

File: modules/buckets/README.md
Comment (SirSpidey) [reply modules/buckets/README.md:61]: Examples section is still here. Is that correct?

**Review by SirSpidey — APPROVED** [review]
Nothing to prevent my approval. However, I've made a couple of comments and have a question about the examples section in the /buckets readme file.

In general, we avoid the future tense (as in `An example that will create...`) and use present tense (`This example creates...`) because it's simpler to parse. But no need to redo this PR.

File: modules/buckets/README.md
Comment (ocofaigh) [reply modules/buckets/README.md:61]: Examples should be removed from here I think

**Review by ocofaigh — APPROVED** [review]

**Pull request merged**
